ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+            }
+        }
+    }
+
+    /* Invoke Mm to return quotas */
+    DPRINT("PspReturnExcessQuotas(): Amount of quota released -- %lu\n", AmountToReturn);
+    MmReturnPoolQuota(QuotaType, AmountToReturn);
+    *ReturnedQuotas = AmountToReturn;
+}
+
+/**
+ * @brief
+ * Internal kernel function that provides the
+ * bulk logic of process quota charging,
+ * necessary for exported kernel routines
+ * needed for quota management.
+ *
+ * @param[in] Process
+ * A process, represented as a EPROCESS object.
+ * This parameter is used to charge the own
+ * process' quota usage.
+ *
+ * @param[in] QuotaBlock
+ * The quota block which quotas are to be charged.
+ * This block can either come from the process itself
+ * or from an object with specified quota charges.
+ *
+ * @param[in] QuotaType
+ * The quota type which quota in question is to
+ * be charged. The permitted types are PsPagedPool,
+ * PsNonPagedPool and PsPageFile.
+ *
+ * @param[in] Amount
+ * The amount of quota to be charged.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if quota charging has
+ * been done successfully without problemns.
+ * STATUS_QUOTA_EXCEEDED is returned if the caller
+ * wants to charge quotas with amount way over
+ * the limits. STATUS_PAGEFILE_QUOTA_EXCEEDED
+ * is returned for the same situation but
+ * specific to page files instead.
  */
 NTSTATUS
 NTAPI
-PspChargeProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN PS_QUOTA_TYPE QuotaType,
-                                   IN SIZE_T    Amount)
+PspChargeProcessQuotaSpecifiedPool(
+    _In_opt_ PEPROCESS Process,
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock,
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _In_ SIZE_T Amount)
 {
     KIRQL OldIrql;
-    ASSERT(Process);
-    ASSERT(Process != PsInitialSystemProcess);
+    SIZE_T UpdatedLimit, FinalLimit, ReturnedQuotas;
+
+    /* Ensure that we get the right quota pool type */
     ASSERT(QuotaType < PsQuotaTypes);
-    ASSERT(Process->QuotaBlock);
 
     /* Guard our quota in a spin lock */
     KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
 
-    if (Process->QuotaUsage[QuotaType] + Amount >
-        Process->QuotaBlock->QuotaEntry[QuotaType].Limit)
+    /* Are we within the bounds of quota limit? */
+    if (QuotaBlock->QuotaEntry[QuotaType].Usage + Amount >
+        QuotaBlock->QuotaEntry[QuotaType].Limit &&
+        QuotaBlock != &PspDefaultQuotaBlock)
     {
-        DPRINT1("Quota exceeded, but ROS will let it slide...\n");
-        KeReleaseSpinLock(&PspQuotaLock, OldIrql);
-        return STATUS_SUCCESS;
-        //return STATUS_QUOTA_EXCEEDED; /* caller raises the exception */
+        /* We aren't... Is this a page file quota charging? */
+        if (QuotaType == PsPageFile)
+        {
+            /* It is, return the appropriate status code */
+            DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Quota amount exceeds the limit on page file quota (the limit -- %lu || the given amount -- %lu)\n",
```

Comment:
Here and in other similar comments you can shorten slightly the dprint by omitting "the" in `(the limit -- ` ... `the given amount`